core/shared/src/main/scala/kantan/csv/HeaderDecoder.scala

```diff
@@ -36,18 +36,24 @@ object HeaderDecoder extends GeneratedHeaderDecoders {
   /** Summons an implicit instance of [[HeaderDecoder]] if one can be found, fails compilation otherwise. */
   def apply[A](implicit ev: HeaderDecoder[A]): HeaderDecoder[A] = macro imp.summon[HeaderDecoder[A]]
 
-  // Horrible code that relies on mutability and assumes that `map` contains the right values, indexed from 0 to length.
-  private[csv] def decoderFromMap(map: Map[String, Int], header: Seq[String]): DecodeResult[Array[Int]] = {
-    val mapping = Array.fill(map.size)(-1)
-
-    header.zipWithIndex.foreach {
-      case (name, i) =>
-        map.get(name).foreach(a => mapping(a) = i)
+  private[csv] def determineRowMappings(requiredHeader: Seq[String], csvHeader: Seq[String]): DecodeResult[Seq[Int]] = {
```

Comment:
This is closer to what I had in mind:

```scala
  private[csv] def determineRowMappings(requiredHeader: Seq[String], csvHeader: Seq[String]): DecodeResult[Seq[Int]] =
    requiredHeader.foldLeft((List.empty[String], List.empty[Int])) {
      case ((missing, found), header) =>
        val index = csvHeader.indexOf(header)

        if(index < 0) (header :: missing, found)
        else (missing, index :: found)
    } match {
      case (missing, _) if missing.nonEmpty =>
        DecodeResult.typeError(s"Missing header(s): ${missing.reverse.mkString(", ")}")
      case (_, mapping) => DecodeResult.success(mapping.reverse)
    }
```

The main difference is that the `Left` and `Right` boxing happens at the very end, which I feel makes the code easier to read - we first find all the errors and all the valid indexes, then return a `Left` if at least one error was found, or a `Right` otherwise.

Reply:
Agreed, being easier to read outweighs the extra concatenations you do by keeping both lists the whole way.